paddle/cinn/ir/ir_schedule_util.cc

```diff
@@ -220,35 +221,115 @@ void ReplaceExpr(Expr* source,
 }
 
 std::vector<int> ValidateFactors(const std::vector<int>& factors,
-                                 int total_extent) {
+                                 int total_extent,
+                                 const ModuleExpr& module_expr) {
+  class NegativeFactorErrorHandler : public IRScheduleErrorHandler {
+   public:
+    explicit NegativeFactorErrorHandler(int64_t factor,
+                                        size_t idx,
+                                        const ModuleExpr& module_expr)
+        : factor_(factor), idx_(idx), module_expr_(module_expr) {}
+
+    std::string GeneralErrorMessage() const final {
+      std::ostringstream os;
+      os << "The params in factors of Split should be positive. However, the "
+            "factor at position "
+         << idx_ << " is " << factor_ << std::endl;
+      return os.str();
+    }
+
+    std::string DetailedErrorMessage() const final {
+      std::ostringstream os;
+      os << GeneralErrorMessage();
+      os << "[Expr info] The Expr of current schedule is: "
+         << module_expr_.GetExprs() << std::endl;
+      return os.str();
+    }
+
+   private:
+    ModuleExpr module_expr_;
+    int64_t factor_;
+    size_t idx_;
+  };
+
+  class InferFactorErrorHandler : public IRScheduleErrorHandler {
+   public:
+    explicit InferFactorErrorHandler(const ModuleExpr& module_expr)
```

Comment:
Will it be complex if we have to implement sub-classes of `IRScheduleErrorHandler` at where we want to report error?

I thought we have a `ErrorHandler` class which can set general error message and detailed message, then we just input strings at where we want to report error.

Reply:
Two reasons:
- Different schedule primitive requires different error message, so using abstract class and sub-classes is to deal with its own need for this.
- Implementation of error message in each subclass is easier to use the macro to simplify the code.